src/System.Slices/System/Buffers/Memory.cs

```diff
@@ -0,0 +1,132 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime;
+
+namespace System.Buffers
+{
+    /// <summary>
+    /// Represents a Span<byte> factory</byte>
+    /// </summary>
+    /// <remarks>
+    /// This struct is not safe for multithreaded access, i.e. it's subject to struct tearing that can result in unsafe memory access. 
+    /// </remarks>
+    public struct Memory<T>
+    {
+        public static unsafe Memory<T> Empty = new Memory<T>(null, 0, 0);
+
+        private readonly T[] _array;
+        private readonly int _offset;
+        private readonly unsafe void* _memory;
+        private readonly int _memoryLength;
+
+        public unsafe Memory(void* pointer, int length)
+        {
+            unsafe
+            {
+                _memory = pointer;
+            }
+
+            _array = null;
+            _offset = 0;
+            _memoryLength = length;
+        }
+
+        public unsafe Memory(T[] array) : this(array, 0, array.Length)
```

Comment:
Is it OK that calling `new Memory<T>(default(T[]));` will cause a NullReferenceException because you have `array.Length`? May be, `array?.Length ?? 0` is better?